File: JAVA/1122. Relative Sort Array.java
class Solution {
    public int[] relativeSortArray(int[] arr1, int[] arr2) {        
         Map<Integer, Integer> countMap = new TreeMap<>();
        for (int num : arr1) {
            countMap.put(num, countMap.getOrDefault(num, 0) + 1);
        }
        List<Integer> output = new ArrayList<>();
        for (int num : arr2) {
            if (countMap.containsKey(num)) {
                int count = countMap.get(num);
                while (count > 0) {
                    output.add(num);
                    count--;
                }
                countMap.remove(num);
            }
        }
        List<Integer> remaining = new ArrayList<>();
        for (int key : countMap.keySet()) {
            int count = countMap.get(key);
            while (count > 0) {
                remaining.add(key);
                count--;
            }
        }
        Collections.sort(remaining);
        output.addAll(remaining);
        int[] ans = new int[output.size()];
        for (int i = 0; i < output.size(); i++) {
           ans[i] = output.get(i);
        }
        return ans;
    }
}

File: C++/1122. Relative Sort Array.cpp
class Solution {
public:
    vector<int> relativeSortArray(vector<int>& arr1, vector<int>& arr2) {
        unordered_map<int, int>hashmap {};
        vector<int>result;
        vector<int>notpresent;
        for(int i = 0; i < arr1.size(); i++){
            hashmap[arr1[i]] += 1;
        }
        for(int j = 0; j < arr2.size(); j++){
            for(int k = 0; k < hashmap[arr2[j]]; k++){
                result.push_back(arr2[j]);
            }
        }
        for(int i = 0; i < arr1.size(); i++){
            auto it = find(arr2.begin(), arr2.end(), arr1[i]);
            bool ispresent = (it != arr2.end());
            if(!ispresent)
                notpresent.push_back(arr1[i]);
        }
        sort(notpresent.begin(),notpresent.end());
        result.insert(result.end(),notpresent.begin(),notpresent.end());
        return result;
    }
};
